pkg/rbd/controllerserver.go

```diff
@@ -220,6 +220,52 @@ func (cs *ControllerServer) checkRBDStatus(rbdVol *rbdVolume, req *csi.CreateVol
 	}
 	return nil
 }
+
+func (cs *ControllerServer) cloneVolumeFromSource(req *csi.CreateVolumeRequest, rbdVol *rbdVolume) error {
+	var err error
+	volumeSource := req.VolumeContentSource
+	// type
+	switch volumeSource.Type.(type) {
+	case *csi.VolumeContentSource_Snapshot:
+		snapName := volumeSource.GetSnapshot().GetSnapshotId()
+		klog.V(2).Infof("creating volume from snapshot %s", snapName)
+		if err = cs.checkSnapshot(req, rbdVol); err != nil {
+			return err
+		}
+	case *csi.VolumeContentSource_Volume:
+		parentVol := volumeSource.GetVolume().GetVolumeId()
+		//check parent volume is present or not
+		//create snapshot
+		//restore snapshot
+		uniqueID := uuid.NewUUID().String()
+		snapName := parentVol + uniqueID
+		snapshotID := "csi-rbd-" + parentVol + "-snap-" + uniqueID
+		//snapName, snapID, VolID string, options, secret
+		rbdSnap, snapErr := cs.doSnapshot(snapName, snapshotID, parentVol, req.GetParameters(), req.GetSecrets())
+		if snapErr != nil {
+			return snapErr
+		}
+
+		err = restoreSnapshot(rbdVol, rbdSnap, rbdVol.AdminID, req.GetSecrets())
+		if err != nil {
+			return err
+		}
+		klog.V(4).Infof("create volume %s from snapshot %s", req.GetName(), rbdSnap.SnapName)
+		// Unprotect snapshot
+		err = unprotectAndDeleteSnapshot(rbdSnap, req.GetSecrets())
```

Comment:
Deleting a snapshot w/ an attached clone is not permitted **unless** your OSDs are configured to only support Mimic-or-later clients. If that is the case, there is no need to protect/unprotect the snapshots.

I actually added the support for simplified image cloning to Mimic to simplify the in-development CSI driver (and avoid the design mistakes of the RBD Cinder driver). I'm perfectly happy w/ this only working for Mimic and later clients, but if that's the case, perhaps we can remove all the protect/unprotect ugliness.

You will also run into issues attempting to delete PVs that have attached clones, which is one of the rationales for adding the "rbd trash rm" command.

Reply:
@dillaman Thanks for the review, i have a few questions


> Deleting a snapshot w/ an attached clone is not permitted **unless** your OSDs are configured to only support Mimic-or-later clients. If that is the case, there is no need to protect/unprotect the snapshots.
> 

is there a way to delete a snapshot w/ an attached clone. from csi perspective, a volume can be deleted if there are snapshots associated to it.

you are suggested that we can don't really need to protect and unprotect the snapshots if osd is configured to support Mimic or latest client?

> I actually added the support for simplified image cloning to Mimic to simplify the in-development CSI driver (and avoid the design mistakes of the RBD Cinder driver). I'm perfectly happy w/ this only working for Mimic and later clients, but if that's the case, perhaps we can remove all the protect/unprotect ugliness.

from CSI point of view, we don't know what is  the backend version is it mimic or latest or even old one (  removing protect and unprotect will cause an issue for the older version?)

> You will also run into issues attempting to delete PVs that have attached clones, which is one of the rationales for adding the "rbd trash rm" command.

yes currently we have an issue for that one i think it is https://github.com/ceph/ceph-csi/issues/227

Reply:
It would be nice if we could just say "must use >= mimic" but that's not really feasible.  For those (like me) that were curious about the special features in mimic, this seemed like good FYI:  https://ceph.com/community/new-mimic-simplified-rbd-image-cloning/

@dillaman I had been looking at using flatten to disassociate volumes that were created from snapshots; and my thought was that works on clones as well (assuming clone via snap).  Curious, is there a concern with that?

Reply:
@Madhu-1 

> is there a way to delete a snapshot w/ an attached clone. from csi perspective, a volume can be deleted if there are snapshots associated to it.

Just to clarify, the spec actually provides some flexibility here and "allows" the deletion of a volume to fail if it has snapshots associated with it.  I suspect though since on RBD snapshots are the clones that's what you were referring to in this case but I just wanted to make sure.  At least in that case we only have to do the flatten here on the "csi clone" or "volume from snap" case.

Reply:
Playing devil's advocate, given that the CSI is so new, is prohibiting cloning for < mimic really that bad? Of course, if you just immediately flatten the clones for the < mimic case, that seems like a fair compromise.

The downside is that (1) any attempts to delete the source PV will fail while the clone is still attached, (2) flattening might take too long (blocking) of an amount of time if it's blocking other PV operations,  (3) you would need to prevent creation of snapshots on the clone until it has been flattened (assuming you could get concurrent PV operations against it).

Reply:
@Madhu-1 
> is there a way to delete a snapshot w/ an attached clone. from csi perspective, a volume can be deleted if there are snapshots associated to it.

This is only possible w/ clone format v2 (requiring a Mimic or later release)

> you are suggested that we can don't really need to protect and unprotect the snapshots if osd is configured to support Mimic or latest client?

Correct -- RBD will automatically use the new clone format v2 if the configured to only support Mimic and later clients. This can be force-overridden by adding ```--rbd_default_clone_format=2``` to the ```rbd clone``` CLI.

> from CSI point of view, we don't know what is the backend version is it mimic or latest or even old one ( removing protect and unprotect will cause an issue for the older version?)

It's trivial to query from the ```ceph``` CLI.

> yes currently we have an issue for that one i think it is #227

Thanks, I'll comment on that.